pkg/controller/volume/persistentvolume/pv_controller.go

```diff
@@ -545,6 +545,19 @@ func (ctrl *PersistentVolumeController) syncVolume(volume *v1.PersistentVolume)
 		if err != nil {
 			return err
 		}
+		if !found {
+			if volume.Status.Phase != v1.VolumeReleased && volume.Status.Phase != v1.VolumeFailed {
+				// If PV is pre-bound by scheduler, it's possible on heavy load
+				// that corresponding PVC is not synced in controller yet. So
+				// we double-check PVC in apiserver to make sure we will not
+				// reclaim a PV due to API delay.
+				obj, err = ctrl.kubeClient.CoreV1().PersistentVolumeClaims(volume.Spec.ClaimRef.Namespace).Get(volume.Spec.ClaimRef.Name, metav1.GetOptions{})
```

Comment:
In what order does scheduler write bound PV and PVC? If it writes PV first, this code might be executed before the scheduler even writes the PVC.

Reply:
Reading #66287, it probably does not matter. Still I'd like to understand the issue before approving this PR.

Reply:
The scheduler only writes PV.ClaimRef.  It doesn't not modify PVC (in the preprovisioned case)

Reply:
I like the idea of only checking this of some special scheduler annotation is set.  Otherwise, this will trigger an extra API call for regular volumes too.

Reply:
Although eventually when almost everything moves to late binding, then it makes no difference at that point.

Reply:
@msau42 
Add a special scheduler `annBoundByScheduler` annotation in:
https://github.com/kubernetes/kubernetes/pull/67062/commits/3508f1effe860dd2ee4bb5ecc3ff8adda5f3e52a.

- use `annBoundByScheduler` to replace `annBoundByController` when PV is bound by scheduler
- check `annBoundByScheduler` and `annBoundByController` annotation both to determine if volume is bound or not
- double-check PVC on apiserver only if PV is bound by scheduler

Reply:
I am not sure about `annBoundByScheduler` - if I have an external PV binder, then I still experience the issue. IMO, it should  just check `annBoundByController` (the controller has seen the PVC so it must be in its cache).

Reply:
Right now the scheduler is already setting annBoundByController.  I need to think about if it is ok to remove that annotation from scheduler.

Reply:
Regardless, I'm not sure how useful it is to add a scheduler annotation because eventually almost everything should be going through the scheduler, and like @jsafrane pointed out, any other external component that is binding PVs could hit this too.

Reply:
I think we cannot remove annBoundByController from the scheduler.  The annotation determines whether or not we automatically try to rollback the binding in case of some error.

Reply:
I see. `Controller` in `annBoundByController` means all possible PV binders, not only PV controller itself. IIUC, if we want to optimize for external PV binders, we need to add a additional annotation for all external PV binders, but that requires upgrading external PV binders.